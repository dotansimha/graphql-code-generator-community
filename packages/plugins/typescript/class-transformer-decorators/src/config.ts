export type ClassTransformerDecoratorConfig = {
  declarationKind?: 'class';
  // List of classes to decorate
  classWhitelist?: string[];
  // RegExp pattern of classes to decorate
  classNamePattern?: string;
};
