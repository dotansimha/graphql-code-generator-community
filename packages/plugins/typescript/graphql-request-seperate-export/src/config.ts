import {
  ClientSideBasePluginConfig,
  RawClientSideBasePluginConfig,
} from '@graphql-codegen/visitor-plugin-common';

export type PluginConfig = {
  /**
   * example: `import { client } from '../your-graphql-client';`
   *
   * the imported `client` should be an instance of graphql-client
   */
  importGraphQLClientStatment: string;

  /**
   * @description By default the `request` method return the `data` or `errors` key from the response. If you need to access the `extensions` key you can use the `rawRequest` method.
   *
   * @exampleMarkdown
   * ```yaml
   * generates:
   * path/to/file.ts:
   *  plugins:
   *    - typescript
   *    - typescript-operations
   *    - typescript-graphql-request-seperate-export
   *  config:
   *    rawRequest: true
   *    pureMagicComment: true
   * ```
   */
  rawRequest?: boolean;
};

/**
 * This plugin generate a some request seperate export function with graphql-request, allow you to easily customize the way you fetch your data, without loosing the strongly-typed integration.
 */
export type RawGraphQLSeperateExportPluginConfig = RawClientSideBasePluginConfig & PluginConfig;
export type GraphQLSeperateExportPluginConfig = ClientSideBasePluginConfig & PluginConfig;
