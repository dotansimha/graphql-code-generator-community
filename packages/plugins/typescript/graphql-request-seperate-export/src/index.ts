import { extname } from 'path';
import { concatAST, FragmentDefinitionNode, GraphQLSchema, Kind } from 'graphql';
import { DocumentNode } from 'graphql';
import { oldVisit, PluginFunction, PluginValidateFn, Types } from '@graphql-codegen/plugin-helpers';
import {
  LoadedFragment,
  RawClientSideBasePluginConfig,
} from '@graphql-codegen/visitor-plugin-common';
import { RawGraphQLSeperateExportPluginConfig } from './config';
import { GraphQLRequestSeperateExportVisitor } from './visitor';

export const plugin: PluginFunction<RawGraphQLSeperateExportPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: RawGraphQLSeperateExportPluginConfig,
) => {
  const allAst = concatAST(documents.map(v => v.document));
  const allFragments: LoadedFragment[] = [
    ...(
      allAst.definitions.filter(
        d => d.kind === Kind.FRAGMENT_DEFINITION,
      ) as FragmentDefinitionNode[]
    ).map(fragmentDef => ({
      node: fragmentDef,
      name: fragmentDef.name.value,
      onType: fragmentDef.typeCondition.name.value,
      isExternal: false,
    })),
    ...(config.externalFragments || []),
  ];
  const visitor = new GraphQLRequestSeperateExportVisitor(schema, allFragments, config);
  const visitorResult = oldVisit(allAst, { leave: visitor }) as DocumentNode;

  return {
    prepend: visitor.getImports(),
    content: [
      visitor.fragments,
      'type GraphQLClientRequestHeaders = Headers | string[][] | Record<string, string>',
      ...visitorResult.definitions.filter(t => typeof t === 'string'),
      visitor.sdkContent,
    ].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  _schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  _config: RawClientSideBasePluginConfig,
  outputFile: string,
) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(
      `Plugin "typescript-graphql-request-seperate-export" requires extension to be ".ts"!`,
    );
  }
};

export { GraphQLRequestSeperateExportVisitor };
