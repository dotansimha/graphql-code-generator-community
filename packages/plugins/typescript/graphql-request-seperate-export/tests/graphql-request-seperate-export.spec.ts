import { buildClientSchema, GraphQLSchema, parse } from 'graphql';
import { mergeOutputs, Types } from '@graphql-codegen/plugin-helpers';
import { validateTs } from '@graphql-codegen/testing';
import { plugin as tsPlugin, TypeScriptPluginConfig } from '@graphql-codegen/typescript';
import {
  plugin as tsDocumentsPlugin,
  TypeScriptDocumentsPluginConfig,
} from '@graphql-codegen/typescript-operations';
import { DocumentMode } from '@graphql-codegen/visitor-plugin-common';
import { RawGraphQLSeperateExportPluginConfig } from '../src/config';
import { plugin } from '../src/index';

const schema = buildClientSchema(require('../../../../../dev-test/githunt/schema.json'));
const basicDoc = parse(/* GraphQL */ `
  query feed {
    feed {
      id
      commentCount
      repository {
        owner {
          avatar_url
        }
      }
    }
  }

  query feed2($v: String!) {
    feed {
      id
    }
  }

  query feed3($v: String) {
    feed {
      id
    }
  }

  query feed4($v: String! = "TEST") {
    feed {
      id
    }
  }
`);

const unnamedDoc = parse(/* GraphQL */ `
  {
    feed {
      id
    }
  }
`);

const validate = async (
  content: Types.PluginOutput,
  config: TypeScriptPluginConfig &
    TypeScriptDocumentsPluginConfig &
    RawGraphQLSeperateExportPluginConfig,
  docs: Types.DocumentFile[],
  pluginSchema: GraphQLSchema,
  usage: string,
) => {
  const m = mergeOutputs([
    await tsPlugin(pluginSchema, docs, config, { outputFile: '' }),
    await tsDocumentsPlugin(pluginSchema, docs, config, { outputFile: '' }),
    content,
    usage,
  ]);

  validateTs(m, {
    allowSyntheticDefaultImports: true,
  });

  return m;
};

const importGraphQLClientStatment = 'import { client } from "./my-graphql-request"';

describe('graphql-request-seperate-export', () => {
  describe('generate split request function', () => {
    it('Should generate a correct wrap method', async () => {
      const config = {
        importGraphQLClientStatment,
        pureMagicComment: true,
      };
      const docs = [{ filePath: '', document: basicDoc }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      const usage = `
async function test() {
  await feed();
  await feed3();
  await feed4();

  const result = await feed2({ v: "1" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}`;
      const output = await validate(result, config, docs, schema, usage);

      expect(output).toMatchSnapshot();
    });

    it('Should generate a correct wrap method with documentMode=string', async () => {
      const config = {
        importGraphQLClientStatment,
        documentMode: DocumentMode.string,
      };
      const docs = [{ filePath: '', document: basicDoc }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      const usage = `
async function test() {
  await feed();
  await feed3();
  await feed4();

  const result = await feed2({ v: "1" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}`;
      const output = await validate(result, config, docs, schema, usage);

      expect(output).toMatchSnapshot();
    });

    it('Should support rawRequest', async () => {
      const config = {
        importGraphQLClientStatment,
        rawRequest: true,
      };
      const docs = [{ filePath: '', document: basicDoc }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      const usage = `
        async function rawRequestTest() {
          await feed();
          await feed3();
          await feed4();

          const result = await feed2({ v: "1" });

          if (result.data.feed) {
            if (result.data.feed[0]) {
              const id = result.data.feed[0].id
            }
          }
        }
      `;
      const output = await validate(result, config, docs, schema, usage);

      expect(output).toMatchSnapshot();
    });

    it('Should throw if it encounters unnamed operations', async () => {
      const config = {
        importGraphQLClientStatment,
      };
      const docs = [{ filePath: '', document: unnamedDoc }];
      try {
        await plugin(schema, docs, config, { outputFile: 'graphql.ts' });
        fail('Should throw');
      } catch (err: unknown) {
        expect(err).toMatchInlineSnapshot(`
          [Error: Plugin 'graphql-request-seperate-export' cannot generate request functions for unnamed operation.

          {
            feed {
              id
            }
          }]
        `);
      }
    });
  });
});
