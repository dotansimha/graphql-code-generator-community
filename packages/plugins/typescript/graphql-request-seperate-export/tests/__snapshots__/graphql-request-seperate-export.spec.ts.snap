// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-request-seperate-export generate split request function Should generate a correct wrap method 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
import { client } from "./my-graphql-request"
import gql from 'graphql-tag';
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  /** A feed of repository submissions */
  feed?: Maybe<Array<Maybe<Entry>>>;
  /** A single entry */
  entry?: Maybe<Entry>;
  /** Return the currently logged in user, or null if nobody is logged in */
  currentUser?: Maybe<User>;
};


export type QueryFeedArgs = {
  type: FeedType;
  offset?: InputMaybe<Scalars['Int']>;
  limit?: InputMaybe<Scalars['Int']>;
};


export type QueryEntryArgs = {
  repoFullName: Scalars['String'];
};

/** A list of options for the sort order of the feed */
export enum FeedType {
  /** Sort by a combination of freshness and score, using Reddit's algorithm */
  Hot = 'HOT',
  /** Newest entries first */
  New = 'NEW',
  /** Highest score entries first */
  Top = 'TOP'
}

/** Information about a GitHub repository submitted to GitHunt */
export type Entry = {
  __typename?: 'Entry';
  /** Information about the repository from GitHub */
  repository: Repository;
  /** The GitHub user who submitted this entry */
  postedBy: User;
  /** A timestamp of when the entry was submitted */
  createdAt: Scalars['Float'];
  /** The score of this repository, upvotes - downvotes */
  score: Scalars['Int'];
  /** The hot score of this repository */
  hotScore: Scalars['Float'];
  /** Comments posted about this repository */
  comments: Array<Maybe<Comment>>;
  /** The number of comments posted about this repository */
  commentCount: Scalars['Int'];
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** XXX to be changed */
  vote: Vote;
};


/** Information about a GitHub repository submitted to GitHunt */
export type EntryCommentsArgs = {
  limit?: InputMaybe<Scalars['Int']>;
  offset?: InputMaybe<Scalars['Int']>;
};

/**
 * A repository object from the GitHub API. This uses the exact field names returned by the
 * GitHub API for simplicity, even though the convention for GraphQL is usually to camel case.
 */
export type Repository = {
  __typename?: 'Repository';
  /** Just the name of the repository, e.g. GitHunt-API */
  name: Scalars['String'];
  /** The full name of the repository with the username, e.g. apollostack/GitHunt-API */
  full_name: Scalars['String'];
  /** The description of the repository */
  description?: Maybe<Scalars['String']>;
  /** The link to the repository on GitHub */
  html_url: Scalars['String'];
  /** The number of people who have starred this repository on GitHub */
  stargazers_count: Scalars['Int'];
  /** The number of open issues on this repository on GitHub */
  open_issues_count?: Maybe<Scalars['Int']>;
  /** The owner of this repository on GitHub, e.g. apollostack */
  owner?: Maybe<User>;
};

/** A user object from the GitHub API. This uses the exact field names returned from the GitHub API. */
export type User = {
  __typename?: 'User';
  /** The name of the user, e.g. apollostack */
  login: Scalars['String'];
  /** The URL to a directly embeddable image for this user's avatar */
  avatar_url: Scalars['String'];
  /** The URL of this user's GitHub page */
  html_url: Scalars['String'];
};

/** A comment about an entry, submitted by a user */
export type Comment = {
  __typename?: 'Comment';
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** The GitHub user who posted the comment */
  postedBy: User;
  /** A timestamp of when the comment was posted */
  createdAt: Scalars['Float'];
  /** The text of the comment */
  content: Scalars['String'];
  /** The repository which this comment is about */
  repoName: Scalars['String'];
};

/** XXX to be removed */
export type Vote = {
  __typename?: 'Vote';
  vote_value: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  /** Submit a new repository, returns the new submission */
  submitRepository?: Maybe<Entry>;
  /** Vote on a repository submission, returns the submission that was voted on */
  vote?: Maybe<Entry>;
  /** Comment on a repository, returns the new comment */
  submitComment?: Maybe<Comment>;
};


export type MutationSubmitRepositoryArgs = {
  repoFullName: Scalars['String'];
};


export type MutationVoteArgs = {
  repoFullName: Scalars['String'];
  type: VoteType;
};


export type MutationSubmitCommentArgs = {
  repoFullName: Scalars['String'];
  commentContent: Scalars['String'];
};

/** The type of vote to record, when submitting a vote */
export enum VoteType {
  Up = 'UP',
  Down = 'DOWN',
  Cancel = 'CANCEL'
}

export type Subscription = {
  __typename?: 'Subscription';
  /** Subscription fires on every comment added */
  commentAdded?: Maybe<Comment>;
};


export type SubscriptionCommentAddedArgs = {
  repoFullName: Scalars['String'];
};
export type FeedQueryVariables = Exact<{ [key: string]: never; }>;


export type FeedQuery = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number, commentCount: number, repository: { __typename?: 'Repository', owner?: { __typename?: 'User', avatar_url: string } | null } } | null> | null };

export type Feed2QueryVariables = Exact<{
  v: Scalars['String'];
}>;


export type Feed2Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

export type Feed3QueryVariables = Exact<{
  v?: InputMaybe<Scalars['String']>;
}>;


export type Feed3Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

export type Feed4QueryVariables = Exact<{
  v?: Scalars['String'];
}>;


export type Feed4Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

type GraphQLClientRequestHeaders = Headers | string[][] | Record<string, string>
export const FeedDocument = /*#__PURE__*/ gql\`
    query feed {
  feed {
    id
    commentCount
    repository {
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const Feed2Document = /*#__PURE__*/ gql\`
    query feed2($v: String!) {
  feed {
    id
  }
}
    \`;
export const Feed3Document = /*#__PURE__*/ gql\`
    query feed3($v: String) {
  feed {
    id
  }
}
    \`;
export const Feed4Document = /*#__PURE__*/ gql\`
    query feed4($v: String! = "TEST") {
  feed {
    id
  }
}
    \`;

export type FeedReturnType = FeedQuery

export const feed = /*#__PURE__*/async (variables?: FeedQueryVariables, options?: GraphQLClientRequestHeaders): Promise<FeedReturnType> => {
  return client.request<FeedQuery, FeedQueryVariables>(FeedDocument, variables || {}, options) as Promise<FeedReturnType>;
}

export type Feed2ReturnType = Feed2Query

export const feed2 = /*#__PURE__*/async (variables: Feed2QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed2ReturnType> => {
  return client.request<Feed2Query, Feed2QueryVariables>(Feed2Document, variables , options) as Promise<Feed2ReturnType>;
}

export type Feed3ReturnType = Feed3Query

export const feed3 = /*#__PURE__*/async (variables?: Feed3QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed3ReturnType> => {
  return client.request<Feed3Query, Feed3QueryVariables>(Feed3Document, variables || {}, options) as Promise<Feed3ReturnType>;
}

export type Feed4ReturnType = Feed4Query

export const feed4 = /*#__PURE__*/async (variables?: Feed4QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed4ReturnType> => {
  return client.request<Feed4Query, Feed4QueryVariables>(Feed4Document, variables || {}, options) as Promise<Feed4ReturnType>;
}
async function test() {
  await feed();
  await feed3();
  await feed4();

  const result = await feed2({ v: "1" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}"
`;

exports[`graphql-request-seperate-export generate split request function Should generate a correct wrap method with documentMode=string 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
import { client } from "./my-graphql-request"
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  /** A feed of repository submissions */
  feed?: Maybe<Array<Maybe<Entry>>>;
  /** A single entry */
  entry?: Maybe<Entry>;
  /** Return the currently logged in user, or null if nobody is logged in */
  currentUser?: Maybe<User>;
};


export type QueryFeedArgs = {
  type: FeedType;
  offset?: InputMaybe<Scalars['Int']>;
  limit?: InputMaybe<Scalars['Int']>;
};


export type QueryEntryArgs = {
  repoFullName: Scalars['String'];
};

/** A list of options for the sort order of the feed */
export enum FeedType {
  /** Sort by a combination of freshness and score, using Reddit's algorithm */
  Hot = 'HOT',
  /** Newest entries first */
  New = 'NEW',
  /** Highest score entries first */
  Top = 'TOP'
}

/** Information about a GitHub repository submitted to GitHunt */
export type Entry = {
  __typename?: 'Entry';
  /** Information about the repository from GitHub */
  repository: Repository;
  /** The GitHub user who submitted this entry */
  postedBy: User;
  /** A timestamp of when the entry was submitted */
  createdAt: Scalars['Float'];
  /** The score of this repository, upvotes - downvotes */
  score: Scalars['Int'];
  /** The hot score of this repository */
  hotScore: Scalars['Float'];
  /** Comments posted about this repository */
  comments: Array<Maybe<Comment>>;
  /** The number of comments posted about this repository */
  commentCount: Scalars['Int'];
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** XXX to be changed */
  vote: Vote;
};


/** Information about a GitHub repository submitted to GitHunt */
export type EntryCommentsArgs = {
  limit?: InputMaybe<Scalars['Int']>;
  offset?: InputMaybe<Scalars['Int']>;
};

/**
 * A repository object from the GitHub API. This uses the exact field names returned by the
 * GitHub API for simplicity, even though the convention for GraphQL is usually to camel case.
 */
export type Repository = {
  __typename?: 'Repository';
  /** Just the name of the repository, e.g. GitHunt-API */
  name: Scalars['String'];
  /** The full name of the repository with the username, e.g. apollostack/GitHunt-API */
  full_name: Scalars['String'];
  /** The description of the repository */
  description?: Maybe<Scalars['String']>;
  /** The link to the repository on GitHub */
  html_url: Scalars['String'];
  /** The number of people who have starred this repository on GitHub */
  stargazers_count: Scalars['Int'];
  /** The number of open issues on this repository on GitHub */
  open_issues_count?: Maybe<Scalars['Int']>;
  /** The owner of this repository on GitHub, e.g. apollostack */
  owner?: Maybe<User>;
};

/** A user object from the GitHub API. This uses the exact field names returned from the GitHub API. */
export type User = {
  __typename?: 'User';
  /** The name of the user, e.g. apollostack */
  login: Scalars['String'];
  /** The URL to a directly embeddable image for this user's avatar */
  avatar_url: Scalars['String'];
  /** The URL of this user's GitHub page */
  html_url: Scalars['String'];
};

/** A comment about an entry, submitted by a user */
export type Comment = {
  __typename?: 'Comment';
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** The GitHub user who posted the comment */
  postedBy: User;
  /** A timestamp of when the comment was posted */
  createdAt: Scalars['Float'];
  /** The text of the comment */
  content: Scalars['String'];
  /** The repository which this comment is about */
  repoName: Scalars['String'];
};

/** XXX to be removed */
export type Vote = {
  __typename?: 'Vote';
  vote_value: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  /** Submit a new repository, returns the new submission */
  submitRepository?: Maybe<Entry>;
  /** Vote on a repository submission, returns the submission that was voted on */
  vote?: Maybe<Entry>;
  /** Comment on a repository, returns the new comment */
  submitComment?: Maybe<Comment>;
};


export type MutationSubmitRepositoryArgs = {
  repoFullName: Scalars['String'];
};


export type MutationVoteArgs = {
  repoFullName: Scalars['String'];
  type: VoteType;
};


export type MutationSubmitCommentArgs = {
  repoFullName: Scalars['String'];
  commentContent: Scalars['String'];
};

/** The type of vote to record, when submitting a vote */
export enum VoteType {
  Up = 'UP',
  Down = 'DOWN',
  Cancel = 'CANCEL'
}

export type Subscription = {
  __typename?: 'Subscription';
  /** Subscription fires on every comment added */
  commentAdded?: Maybe<Comment>;
};


export type SubscriptionCommentAddedArgs = {
  repoFullName: Scalars['String'];
};
export type FeedQueryVariables = Exact<{ [key: string]: never; }>;


export type FeedQuery = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number, commentCount: number, repository: { __typename?: 'Repository', owner?: { __typename?: 'User', avatar_url: string } | null } } | null> | null };

export type Feed2QueryVariables = Exact<{
  v: Scalars['String'];
}>;


export type Feed2Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

export type Feed3QueryVariables = Exact<{
  v?: InputMaybe<Scalars['String']>;
}>;


export type Feed3Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

export type Feed4QueryVariables = Exact<{
  v?: Scalars['String'];
}>;


export type Feed4Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

type GraphQLClientRequestHeaders = Headers | string[][] | Record<string, string>
export const FeedDocument = \`
    query feed {
  feed {
    id
    commentCount
    repository {
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const Feed2Document = \`
    query feed2($v: String!) {
  feed {
    id
  }
}
    \`;
export const Feed3Document = \`
    query feed3($v: String) {
  feed {
    id
  }
}
    \`;
export const Feed4Document = \`
    query feed4($v: String! = "TEST") {
  feed {
    id
  }
}
    \`;

export type FeedReturnType = FeedQuery

export const feed = async (variables?: FeedQueryVariables, options?: GraphQLClientRequestHeaders): Promise<FeedReturnType> => {
  return client.request<FeedQuery, FeedQueryVariables>(FeedDocument, variables || {}, options) as Promise<FeedReturnType>;
}

export type Feed2ReturnType = Feed2Query

export const feed2 = async (variables: Feed2QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed2ReturnType> => {
  return client.request<Feed2Query, Feed2QueryVariables>(Feed2Document, variables , options) as Promise<Feed2ReturnType>;
}

export type Feed3ReturnType = Feed3Query

export const feed3 = async (variables?: Feed3QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed3ReturnType> => {
  return client.request<Feed3Query, Feed3QueryVariables>(Feed3Document, variables || {}, options) as Promise<Feed3ReturnType>;
}

export type Feed4ReturnType = Feed4Query

export const feed4 = async (variables?: Feed4QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed4ReturnType> => {
  return client.request<Feed4Query, Feed4QueryVariables>(Feed4Document, variables || {}, options) as Promise<Feed4ReturnType>;
}
async function test() {
  await feed();
  await feed3();
  await feed4();

  const result = await feed2({ v: "1" });

  if (result.feed) {
    if (result.feed[0]) {
      const id = result.feed[0].id
    }
  }
}"
`;

exports[`graphql-request-seperate-export generate split request function Should support rawRequest 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
import { client } from "./my-graphql-request"
import gql from 'graphql-tag';
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type Query = {
  __typename?: 'Query';
  /** A feed of repository submissions */
  feed?: Maybe<Array<Maybe<Entry>>>;
  /** A single entry */
  entry?: Maybe<Entry>;
  /** Return the currently logged in user, or null if nobody is logged in */
  currentUser?: Maybe<User>;
};


export type QueryFeedArgs = {
  type: FeedType;
  offset?: InputMaybe<Scalars['Int']>;
  limit?: InputMaybe<Scalars['Int']>;
};


export type QueryEntryArgs = {
  repoFullName: Scalars['String'];
};

/** A list of options for the sort order of the feed */
export enum FeedType {
  /** Sort by a combination of freshness and score, using Reddit's algorithm */
  Hot = 'HOT',
  /** Newest entries first */
  New = 'NEW',
  /** Highest score entries first */
  Top = 'TOP'
}

/** Information about a GitHub repository submitted to GitHunt */
export type Entry = {
  __typename?: 'Entry';
  /** Information about the repository from GitHub */
  repository: Repository;
  /** The GitHub user who submitted this entry */
  postedBy: User;
  /** A timestamp of when the entry was submitted */
  createdAt: Scalars['Float'];
  /** The score of this repository, upvotes - downvotes */
  score: Scalars['Int'];
  /** The hot score of this repository */
  hotScore: Scalars['Float'];
  /** Comments posted about this repository */
  comments: Array<Maybe<Comment>>;
  /** The number of comments posted about this repository */
  commentCount: Scalars['Int'];
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** XXX to be changed */
  vote: Vote;
};


/** Information about a GitHub repository submitted to GitHunt */
export type EntryCommentsArgs = {
  limit?: InputMaybe<Scalars['Int']>;
  offset?: InputMaybe<Scalars['Int']>;
};

/**
 * A repository object from the GitHub API. This uses the exact field names returned by the
 * GitHub API for simplicity, even though the convention for GraphQL is usually to camel case.
 */
export type Repository = {
  __typename?: 'Repository';
  /** Just the name of the repository, e.g. GitHunt-API */
  name: Scalars['String'];
  /** The full name of the repository with the username, e.g. apollostack/GitHunt-API */
  full_name: Scalars['String'];
  /** The description of the repository */
  description?: Maybe<Scalars['String']>;
  /** The link to the repository on GitHub */
  html_url: Scalars['String'];
  /** The number of people who have starred this repository on GitHub */
  stargazers_count: Scalars['Int'];
  /** The number of open issues on this repository on GitHub */
  open_issues_count?: Maybe<Scalars['Int']>;
  /** The owner of this repository on GitHub, e.g. apollostack */
  owner?: Maybe<User>;
};

/** A user object from the GitHub API. This uses the exact field names returned from the GitHub API. */
export type User = {
  __typename?: 'User';
  /** The name of the user, e.g. apollostack */
  login: Scalars['String'];
  /** The URL to a directly embeddable image for this user's avatar */
  avatar_url: Scalars['String'];
  /** The URL of this user's GitHub page */
  html_url: Scalars['String'];
};

/** A comment about an entry, submitted by a user */
export type Comment = {
  __typename?: 'Comment';
  /** The SQL ID of this entry */
  id: Scalars['Int'];
  /** The GitHub user who posted the comment */
  postedBy: User;
  /** A timestamp of when the comment was posted */
  createdAt: Scalars['Float'];
  /** The text of the comment */
  content: Scalars['String'];
  /** The repository which this comment is about */
  repoName: Scalars['String'];
};

/** XXX to be removed */
export type Vote = {
  __typename?: 'Vote';
  vote_value: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  /** Submit a new repository, returns the new submission */
  submitRepository?: Maybe<Entry>;
  /** Vote on a repository submission, returns the submission that was voted on */
  vote?: Maybe<Entry>;
  /** Comment on a repository, returns the new comment */
  submitComment?: Maybe<Comment>;
};


export type MutationSubmitRepositoryArgs = {
  repoFullName: Scalars['String'];
};


export type MutationVoteArgs = {
  repoFullName: Scalars['String'];
  type: VoteType;
};


export type MutationSubmitCommentArgs = {
  repoFullName: Scalars['String'];
  commentContent: Scalars['String'];
};

/** The type of vote to record, when submitting a vote */
export enum VoteType {
  Up = 'UP',
  Down = 'DOWN',
  Cancel = 'CANCEL'
}

export type Subscription = {
  __typename?: 'Subscription';
  /** Subscription fires on every comment added */
  commentAdded?: Maybe<Comment>;
};


export type SubscriptionCommentAddedArgs = {
  repoFullName: Scalars['String'];
};
export type FeedQueryVariables = Exact<{ [key: string]: never; }>;


export type FeedQuery = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number, commentCount: number, repository: { __typename?: 'Repository', owner?: { __typename?: 'User', avatar_url: string } | null } } | null> | null };

export type Feed2QueryVariables = Exact<{
  v: Scalars['String'];
}>;


export type Feed2Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

export type Feed3QueryVariables = Exact<{
  v?: InputMaybe<Scalars['String']>;
}>;


export type Feed3Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

export type Feed4QueryVariables = Exact<{
  v?: Scalars['String'];
}>;


export type Feed4Query = { __typename?: 'Query', feed?: Array<{ __typename?: 'Entry', id: number } | null> | null };

type GraphQLClientRequestHeaders = Headers | string[][] | Record<string, string>
export const FeedDocument = gql\`
    query feed {
  feed {
    id
    commentCount
    repository {
      owner {
        avatar_url
      }
    }
  }
}
    \`;
export const Feed2Document = gql\`
    query feed2($v: String!) {
  feed {
    id
  }
}
    \`;
export const Feed3Document = gql\`
    query feed3($v: String) {
  feed {
    id
  }
}
    \`;
export const Feed4Document = gql\`
    query feed4($v: String! = "TEST") {
  feed {
    id
  }
}
    \`;

export type FeedReturnType = { data?: FeedQuery, errors?: Array<{ message: string; extensions?: unknown }>, extensions?: unknown }

export const feed = async (variables?: FeedQueryVariables, options?: GraphQLClientRequestHeaders): Promise<FeedReturnType> => {
  return client.rawRequest<FeedQuery, FeedQueryVariables>(FeedDocument, variables || {}, options) as Promise<FeedReturnType>;
}

export type Feed2ReturnType = { data?: Feed2Query, errors?: Array<{ message: string; extensions?: unknown }>, extensions?: unknown }

export const feed2 = async (variables: Feed2QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed2ReturnType> => {
  return client.rawRequest<Feed2Query, Feed2QueryVariables>(Feed2Document, variables , options) as Promise<Feed2ReturnType>;
}

export type Feed3ReturnType = { data?: Feed3Query, errors?: Array<{ message: string; extensions?: unknown }>, extensions?: unknown }

export const feed3 = async (variables?: Feed3QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed3ReturnType> => {
  return client.rawRequest<Feed3Query, Feed3QueryVariables>(Feed3Document, variables || {}, options) as Promise<Feed3ReturnType>;
}

export type Feed4ReturnType = { data?: Feed4Query, errors?: Array<{ message: string; extensions?: unknown }>, extensions?: unknown }

export const feed4 = async (variables?: Feed4QueryVariables, options?: GraphQLClientRequestHeaders): Promise<Feed4ReturnType> => {
  return client.rawRequest<Feed4Query, Feed4QueryVariables>(Feed4Document, variables || {}, options) as Promise<Feed4ReturnType>;
}
        async function rawRequestTest() {
          await feed();
          await feed3();
          await feed4();

          const result = await feed2({ v: "1" });

          if (result.data.feed) {
            if (result.data.feed[0]) {
              const id = result.data.feed[0].id
            }
          }
        }
      "
`;
